cip.go

```diff
@@ -108,10 +109,30 @@ func main() {
 	manifestBasedSnapshotOf := flag.String(
 		"manifest-based-snapshot-of",
 		"",
-		"read all images in either -manifest or -manifest-dir and print all images that will be promoted to the given registry; this is like -snapshot, but instead of reading from a registry, it reads from the manifests those images that need to be promoted to the given registry")
+		"read all images in either -manifest or -manifest-dir and print all images that should be promoted to the given registry (assuming the given registry is empty); this is like -snapshot, but instead of reading over the network from a registry, it reads from the local manifests only")
 	useServiceAccount := false
 	flag.BoolVar(&useServiceAccount, "use-service-account", false,
 		"pass '--account=...' to all gcloud calls (default: false)")
+	auditorPtr := flag.Bool(
```

Comment:
Without knowing this code very well - I'd expect the audit-mode to be roughly the same as for non-audit.  Trigger, read manifests, do work (alert if audit, else promote).

I guess non-audit mode reals files directly and audit mode reads the URL - is that why they are different?  Would it make sense (medium term?) to converge those?

Reply:
The auditing code shares some code with non-audit code, but the behavior is different:

- non-audit code: read manifests on disk, perform GCR promotions (image copies) using go-containerregistry API
- audit code: read manifests by URL, read HTTP req with GCR state change, check whether state change is pre-approved by the manifests, alert (Stackdriver Error Reporting API) if change is not verifiable

The convergence of the "read manifests" bit at the start would be tricky, mainly because of the way the non-audit code works in harmony with Github PRs (which could have manifests on different branches). I don't think it's worth the effort to converge them.